### PR TITLE
Functions as first-class types.

### PR DESCRIPTION
**Pull request by chriseth**
This adds function types to the types available to the user. There are internal and external functions, the former of which cannot be used in external function calls (one of the reasons is that it would not be possible to evaluate them).

Fixes https://github.com/ethereum/solidity/issues/910, #1353.

Still to do:
- [x] documentation
- [x] discussion if this is the syntax we want
- [x] discussion if `internal` should be the default for function types
- [x] discussion if that is the encoding we want (left aligned `<function id><address>`)
- [x] special case in parser for function types for storage variables (distinguish from fallback function)
- [x] storing function types in storage (and tests for that)
- [x] arrays of functions (also in external function calls)
- [x] squash the commits
- [x] check that delete works on function types (in storage and on stack)
- [x] internal functions should survive the transition from constructor to runtime context in storage.
- [x] zero value for storage function should throw (for internal function, this is a problem because zero is actually a valid jumpdest) solution: invert the value (better than e.g. adding 1 because it can be equally applied to construction and runtime value)
- [x] modify the grammar file
- [x] use new proper ABI type for functions

### CODE REVIEW

**Comment by chriseth** [comment]
This PR already provides basic functionality, you can pass along internal functions in internal function calls and evaluate them and you can pass along external functions in external function calls.

**Comment by chriseth** [comment]
One suggestion concerning notation: TypeScript would use `(uint, uint) => (uint)` for a function type. It would not be too obvious where to place the `external` / `internal` keyword, but it would be much shorter.

**Comment by VoR0220** [comment]
agreed with the above suggestion as it also makes it more readable and able to discern which function is the "root" function and which are the parameters or in scope functions.

**Comment by axic** [comment]
>  `function eval(function(uint) external returns (uint) x, uint a) returns (uint) {`

I think we should come up with a better way of representing this, maybe requiring parentheses around could work: `(function(uint) external returns (uint)) x`

**Comment by chriseth** [comment]
Parentheses might improve readability but it is not too logical (especially consider the two cases where you have only the type and where you also have a variable name), so I would not force the requirement.

**Comment by axic** [comment]
Actually this is the time where the `alias` / `typedef` keyword could be useful :smiley:

**Comment by VoR0220** [comment]
I think `function eval(function(uint) external => (uint) x, uint a) returns (uint) {` would be better....the arrow makes it much more readable...perhaps some enforcement by the compiler that function types go at the end of the args? Or at the very least put that in the style guide...

**Comment by pirapira** [comment]
@VoR0220 ,  @chriseth and I were discussing if the keywords `external` `internal` `pure` or `view` should come before/after the arrow `=>`.  Things before the arrow should talk about the input, and things after the arrow should talk about the output, but `external` talks about neither.  Maybe `external` and `internal` should come before `function`, so:

```
function eval(external function(uint) => (uint) x, uint a) returns (uint) {
              ^^^^^^^^
```

**Comment by VoR0220** [comment]
yeah but then you have to think about how that syncs with the rest of language...hmmmm...

**Comment by axic** [comment]
> Parentheses might improve readability but it is not too logical (especially consider the two cases where you have only the type and where you also have a variable name)

`(function())` and `(function()) x`

I don't think it looks that horrible. (Could be better.) The aim is to have the same syntax as current function definitions.

> so I would not force the requirement.

Even if it is not enforced, but is accepted by the parser, it could go into the style guide.

**Comment by axic** [comment]
>  Maybe `external` and `internal` should come before function

I like this as a future change, which applies to actual function definitions too. However, where do you put `payable`, `view`, `pure` and the modifiers?

**Comment by pirapira** [comment]
@axic `payable` `view` and `pure` can be in the same position as `external` `internal` i.e. before `function`.  I doubt we should include general modifiers in the type of functions.

**Comment by axic** [comment]
> I doubt we should include general modifiers in the type of functions.

I'm thinking about using the same notation (whatever we come up with) for both declarations and variables.

**Comment by pirapira** [comment]
@axic When you receive a function `g` in `function f (function g() onlyOwner)` how do we check whether `g` behaves as `onlyOwner`?  Do we always have access to the Solidity source?

**Comment by axic** [comment]
@pirapira sorry, the quote was misleading. I don't think modifiers apply to function types. However I would like to use the same notation in general (with the exception that modifiers, `public` and `private` are not valid for types).

**Comment by pirapira** [comment]
@axic I agree that we should make the notations uniform here and in function declarations.

File: libsolidity/ast/ASTJsonConverter.cpp [hunk 1]
```diff
@@ -225,6 +225,15 @@ bool ASTJsonConverter::visit(UserDefinedTypeName const& _node)
 	return true;
 }
 
+bool ASTJsonConverter::visit(FunctionTypeName const& _node)
+{
+	addJsonNode(_node, "FunctionTypeName", {
+		make_pair("payable", _node.isPayable()),
+		make_pair("constant", _node.isDeclaredConst())
```
Comment (pirapira) [inline libsolidity/ast/ASTJsonConverter.cpp:238]: Maybe visibility can be added here.

File: libsolidity/codegen/CompilerUtils.cpp [hunk 2]
```diff
@@ -657,6 +683,14 @@ void CompilerUtils::convertType(Type const& _typeOnStack, Type const& _targetTyp
 
 void CompilerUtils::pushZeroValue(Type const& _type)
 {
+	if (auto const* funType = dynamic_cast<FunctionType const*>(&_type))
+	{
+		if (funType->location() == FunctionType::Location::Internal)
+		{
+			m_context << m_context.errorTag();
```
Comment (pirapira) [inline libsolidity/codegen/CompilerUtils.cpp:739]: Why is this a runtime error, not an error reported at the compilation time?

Comment (pirapira) [reply libsolidity/codegen/CompilerUtils.cpp:739]: Ah, now I understand, you need to initialize array elements that contain internal functions, with some values.

File: libsolidity/ast/Types.cpp [hunk 3]
```diff
@@ -1961,6 +2008,26 @@ MemberList::MemberMap FunctionType::nativeMembers(ContractDefinition const*) con
...
+	else
+		return TypePointer();
+}
+
+TypePointer FunctionType::interfaceType(bool /*_inLibrary*/) const
+{
+	if (m_location != Location::External && m_location != Location::Internal)
+		return TypePointer();
+	if (m_location != Location::External)
+		return TypePointer();
```
Comment (pirapira) [inline libsolidity/ast/Types.cpp]: The two checks above look redundant (the first one catches a subset of what the second one catches).

**Review by pirapira — DISMISSED** [review]

File: libsolidity/codegen/CompilerUtils.cpp (hunk 2 above)
Comment (chriseth) [reply libsolidity/codegen/CompilerUtils.cpp:739]: Yes, this is also used when you do `function() x; x();`

File: test/libsolidity/SolidityEndToEndTest.cpp [hunk 4]
```diff
@@ -7366,6 +7366,75 @@ BOOST_AUTO_TEST_CASE(function_type_library_internal)
...
+				return x()()()();
+			}
+		}
+	)";
+
+	compileAndRun(sourceCode, 0, "C");
+	BOOST_CHECK(callContractFunction("f()") == encodeArgs(u256(2)));
+}
+
+BOOST_AUTO_TEST_CASE(array_of_functions)
```
Comment (chriseth) [inline test/libsolidity/SolidityEndToEndTest.cpp]: It seems to be a mapping, not an array

File: test/libsolidity/SolidityEndToEndTest.cpp [hunk 5]
```diff
@@ -7366,6 +7366,75 @@ BOOST_AUTO_TEST_CASE(function_type_library_internal)
...
+	compileAndRun(sourceCode, 0, "C");
+	BOOST_CHECK(callContractFunction("checkSuccess()") == encodeArgs(false));
+	callContractFunction("f()");
+	callContractFunction("f()");
+	BOOST_CHECK(callContractFunction("checkSuccess()") == encodeArgs(false));
+	callContractFunction("f()");
+	BOOST_CHECK(callContractFunction("checkSuccess()") == encodeArgs(true));
+}
+
 // TODO: arrays, libraries with external functions
```
Comment (chriseth) [inline test/libsolidity/SolidityEndToEndTest.cpp]: We should also test that `var x = new function()[](20); x[2]();` correctly throws.

File: test/libsolidity/SolidityEndToEndTest.cpp [hunk 6]
```diff
@@ -7366,6 +7366,75 @@ BOOST_AUTO_TEST_CASE(function_type_library_internal)
...
+
+	compileAndRun(sourceCode, 0, "C");
+	BOOST_CHECK(callContractFunction("f()") == encodeArgs(u256(2)));
+}
+
+BOOST_AUTO_TEST_CASE(array_of_functions)
+{
+	char const* sourceCode = R"(
+		contract Flow {
+			bool success;
```
Comment (chriseth) [inline test/libsolidity/SolidityEndToEndTest.cpp]: Use `bool public success`, then you do not need `checkSuccess`

File: test/libsolidity/SolidityEndToEndTest.cpp [hunk 7]
```diff
@@ -7366,6 +7366,75 @@ BOOST_AUTO_TEST_CASE(function_type_library_internal)
...
+			function stage1() internal {
+					stages[msg.sender] = stage2;
+			}
+
+			function stage2() internal {
+				success = true;
+			}
+
+			function f () {
+				if (0 == steps[msg.sender])
```
Comment (chriseth) [inline test/libsolidity/SolidityEndToEndTest.cpp]: `steps` seems to be a typo.

File: test/libsolidity/SolidityEndToEndTest.cpp (hunk 5 above)
Comment (chriseth) [reply test/libsolidity/SolidityEndToEndTest.cpp]: Oh and do we support comparison between functions? I think we should.

File: test/libsolidity/SolidityEndToEndTest.cpp [hunk 8]
```diff
@@ -7366,6 +7366,75 @@ BOOST_AUTO_TEST_CASE(function_type_library_internal)
...
+					stages[msg.sender] = stage0;
+				stages[msg.sender]();
+			}
+		}
+	)";
+
+	compileAndRun(sourceCode, 0, "C");
+	BOOST_CHECK(callContractFunction("checkSuccess()") == encodeArgs(false));
+	callContractFunction("f()");
```
Comment (chriseth) [inline test/libsolidity/SolidityEndToEndTest.cpp]: Please check the return value.

File: test/libsolidity/SolidityParser.cpp [hunk 9]
```diff
@@ -1311,6 +1311,31 @@ BOOST_AUTO_TEST_CASE(function_type_as_parameter)
...
+				function() returns(function() returns(function() returns(function() returns(uint)))) x;
+				uint y;
+				y = x()()()();
+			}
+		}
+	)";
+	BOOST_CHECK(successParse(text));
+}
+
+BOOST_AUTO_TEST_CASE(array_of_functions)
```
Comment (chriseth) [inline test/libsolidity/SolidityParser.cpp]: Mapping

File: libsolidity/codegen/CompilerUtils.cpp (hunk 2 above)
Comment (pirapira) [reply libsolidity/codegen/CompilerUtils.cpp:739]: In my taste I would disallow `function() x` without an initialization.  Also, in my taste I would disallow array extension without giving elements, but I realized that's a bigger change and for now we allow array extension without giving elements.

File: test/libsolidity/SolidityEndToEndTest.cpp (hunk 6 above)
Comment (pirapira) [reply test/libsolidity/SolidityEndToEndTest.cpp]: I was lazy to look up the name of the accessor function.

File: test/libsolidity/SolidityEndToEndTest.cpp (hunk 8 above)
Comment (pirapira) [reply test/libsolidity/SolidityEndToEndTest.cpp]: I was lazy to lookup what I should expect as the return value from a function that does not return a value.

Comment (chriseth) [reply test/libsolidity/SolidityEndToEndTest.cpp]: `BOOST_CHECK(callContractFunction("f()") == encodeArgs())`

File: test/libsolidity/SolidityEndToEndTest.cpp (hunk 6 above)
Comment (chriseth) [reply test/libsolidity/SolidityEndToEndTest.cpp]: it's just `success()`

File: libsolidity/ast/Types.cpp (hunk 3 above)
Comment (chriseth) [reply libsolidity/ast/Types.cpp]: Should be fixed now.

**Comment by chriseth** [comment]
In the meantime, I assume that we at least settled for the basic notation for function types with the option to also introduce a short notation in the future. I also think that the default value of "internal" for function types is fine.

During testing, I found an unfortunate situation I would like to get opinions on:

```
contract c {
  function (uint) internal returns (uint) x = a;
  function a(uint) internal returns (uint) { return 2; }
}
```

The assignment `x=a` will be done at construction time, but the value of the identifier `a` is different at construction time than it is at runtime (because the code is different). I don't see a real language-based solution that this problem, so we should find a way to access internal functions at construction time that are also valid at runtime. We know which of the internal functions are accessed, so it might make sense to create a "trampoline" at a fixed point which has a special flag so it will not be moved by the optimizer.

**Comment by pirapira** [comment]
@chriseth you can use `CODESIZE - 100` to access the same function at the construction time and at runtime in a uniform way.

**Comment by chriseth** [comment]
@pirapira proposed to represent internal functions as a pair of `(construction_time_label, runtime_label)`. Both can be limited to 4 bytes. If a function is accessed at runtime, it is fine to leave hte first part as zero. If it is accessed at construction time, both are filled. If a function is called at runtime, the upper bits are zeroed out, if it is called at construction time, the lower bits are zeroed out.

File: docs/types.rst [hunk 10]
```diff
@@ -266,6 +266,118 @@ to and from all integer types but implicit conversion is not allowed.
...
     }
 
+.. index:: ! function type, ! type; function
+
+.. _function_types:
+
+Function Types
+--------------
+
+Functions can be assigned to variables and passed on together with function calls.
```
Comment (pirapira) [inline docs/types.rst]: `passed on together with` -> `passed as arguments to`

File: docs/types.rst [hunk 11]
```diff
@@ -266,6 +266,118 @@ to and from all integer types but implicit conversion is not allowed.
...
+These types come in two flavours: *internal* and *external* functions.
+
+Internal functions can only be used inside the current contract (more specifically,
+inside the current code unit, which also includes internal library functions)
+because they cannot be executed outside of the
+context of the current function. Calling an internal function is realized
```
Comment (pirapira) [inline docs/types.rst]: `outside of the context of the current function` -> `outside of the context of the current contract`

File: docs/types.rst [hunk 12]
```diff
@@ -266,6 +266,118 @@ to and from all integer types but implicit conversion is not allowed.
...
+by jumping to its entry label, just like when calling an function of the current
+contract internally.
+
+External functions consist of an address and a function signature and they can
+be passed via and returned from external function calls.
+
+Function types are notated as follows:
+
+    function (<argument types>) internal / external returns (<return types>)
```
Comment (pirapira) [inline docs/types.rst]: Maybe `{internal|external}` to be explicit about what can be altered.

File: docs/types.rst [hunk 13]
```diff
@@ -266,6 +266,118 @@ to and from all integer types but implicit conversion is not allowed.
...
+
+As always, the ``returns (<return types>)`` is optional.
```
Comment (pirapira) [inline docs/types.rst]: What is the interpretation when `returns` is missing?

File: docs/types.rst [hunk 14]
```diff
@@ -266,6 +266,118 @@ to and from all integer types but implicit conversion is not allowed.
...
+          r[i] = f(self[i]);
+        }
+      }
+      function reduce(
+        uint[] memory self,
+        function (uint) returns (uint) f
+      )
+        returns (uint r)
+      {
+        r = self[0];
```
Comment (pirapira) [inline docs/types.rst:334]: Maybe check the length of `self` before accessing an element.

File: docs/types.rst (hunk 10 above)
Comment (pirapira) [reply docs/types.rst]: I want another sentence here.  `The variables and the function arguments that contain functions have function types.  A function type specifies the argument types and the return type of the function.  Moreover (continues next line)`

Comment (chriseth) [reply docs/types.rst]: Agreed that this needs more explanation. I modified your sentence slightly.

File: docs/types.rst (hunk 14 above)
Comment (chriseth) [reply docs/types.rst:334]: It will throw an exception if the array is too short.

**Comment by pirapira** [comment]
Currently the build is broken for me (at commit 59e8c63f183e0f62e9b45049c1b0157d4f5f2ce8).

File: libsolidity/codegen/Compiler.h [hunk 15]
```diff
@@ -35,7 +35,9 @@ class Compiler
 public:
 	explicit Compiler(bool _optimize = false, unsigned _runs = 200):
 		m_optimize(_optimize),
-		m_optimizeRuns(_runs)
+		m_optimizeRuns(_runs),
+		m_context(CompilationMode::Creation, &m_runtimeContext),
```
Comment (chriseth) [inline libsolidity/codegen/Compiler.h]: Not sure it is a good idea to take a reference of a not yet initialized variable.

File: libsolidity/codegen/CompilerContext.h [hunk 16]
```diff
@@ -37,13 +37,24 @@ namespace dev {
...
+enum class CompilationMode { Runtime, Creation };
+
 /**
  * Context to be shared by all units that compile the same contract.
  * It stores the generated bytecode and the position of identifiers in memory and on the stack.
  */
 class CompilerContext
 {
 public:
+	CompilerContext(CompilationMode _mode, CompilerContext* _runtimeContext = nullptr) :
```
Comment (chriseth) [inline libsolidity/codegen/CompilerContext.h]: A less explicit way of doing the same thing would be to use the null-ness of `_runtimeContext` as the switch.

File: libsolidity/codegen/ContractCompiler.h [hunk 17]
```diff
@@ -38,11 +38,11 @@ namespace solidity {
 class ContractCompiler: private ASTConstVisitor
 {
 public:
-	explicit ContractCompiler(CompilerContext& _context, bool _optimise):
+	explicit ContractCompiler(CompilationMode _mode, CompilerContext* _runtimeContext, CompilerContext& _context, bool _optimise):
 		m_optimise(_optimise),
 		m_context(_context)
 	{
-		m_context = CompilerContext();
```
Comment (chriseth) [inline libsolidity/codegen/ContractCompiler.h:45]: I know this is not part of this change, but what is the purpose of `_context` if it is overwritten in `m_context`?

Comment (pirapira) [reply libsolidity/codegen/ContractCompiler.h:45]: This assignment has a side-effect similar to `_context = CompilerContext()`.  The effect is visible from outside of this class.

File: libsolidity/codegen/Compiler.h (hunk 15 above)
Comment (pirapira) [reply libsolidity/codegen/Compiler.h]: I'm assuming that the object is not seen before the constructor finishes.

File: libsolidity/codegen/CompilerContext.h (hunk 16 above)
Comment (pirapira) [reply libsolidity/codegen/CompilerContext.h]: I chose to have redundancy so that it's more difficult to misuse the class.

File: libevmasm/Assembly.cpp [hunk 18]
```diff
@@ -485,6 +485,7 @@ LinkerObject const& Assembly::assemble() const
 			break;
 		case Tag:
 			tagPos[(unsigned)i.data()] = ret.bytecode.size();
+			assertThrow(ret.bytecode.size() < 0xffffffffL, AssemblyException, "Tag too large.");
```
Comment (axic) [inline libevmasm/Assembly.cpp]: Can we pull this into a separate PR and merge it now?

File: libsolidity/codegen/ContractCompiler.cpp [hunk 19]
```diff
@@ -293,6 +293,7 @@ void ContractCompiler::appendCalldataUnpacker(TypePointers const& _typeParameter
 	{
 		// stack: v1 v2 ... v(k-1) base_offset current_offset
 		TypePointer type = parameterType->decodingType();
+		solAssert(type, "No decoding type found.");
```
Comment (axic) [inline libsolidity/codegen/ContractCompiler.cpp:304]: Can we also pull this with other assert?

**Comment by axic** [comment]
Wondering if `payable` should be part of the signature or not.

File: libevmasm/Assembly.cpp (hunk 18 above)
Comment (chriseth) [reply libevmasm/Assembly.cpp]: I don't see it as so critical.

File: libsolidity/codegen/ContractCompiler.cpp (hunk 19 above)
Comment (chriseth) [reply libsolidity/codegen/ContractCompiler.cpp:304]: I'm planning to merge this today or at latest tomorrow.

**Comment by chriseth** [comment]
@axic yes, I think payable already is part of the signature, I just forgot to mention it in the documentation and the grammar.

File: libevmasm/Assembly.cpp (hunk 18 above)
Comment (axic) [reply libevmasm/Assembly.cpp]: You don't see 16 Mb bytecodes flying around? ⛑

**Comment by axic** [comment]
>  discussion if that is the encoding we want (left aligned `<function id><address>`)

Will we also add conversion from `function` type to `address` since that seems natural? A recipient of a function type could decide not to call it if it refers to a third party address for example.

File: libsolidity/codegen/ContractCompiler.cpp (hunk 19 above)
Comment (axic) [reply libsolidity/codegen/ContractCompiler.cpp:304]: Are all the tickboxes done? (See on the top.)

**Comment by axic** [comment]
One shortcoming I see is a limitation we have in the ABI in general: return types are not part of the signature. I would like that to be addressed in the future.

**Comment by axic** [comment]
> By default, function types are internal, so the ``internal`` keyword can be omitted.

In the oracle example, what happens if an ``internal`` function is passed around?

File: docs/types.rst [hunk 20]
```diff
@@ -266,6 +266,122 @@ to and from all integer types but implicit conversion is not allowed.
...
+
+Function types are notated as follows:
+
+    function (<parameter types>) {internal|external} [constant] [returns (<return types>)]
+
+In contrast to the parameter types, the return types cannot be empty - if the
+function type should not return anything, the whole ``returns (<return types>)``
+part has to be omitted.
+
+By default, function types are internal, so the ``internal`` keyword can be
```
Comment (axic) [inline docs/types.rst:300]: In the oracle example, what happens if an `internal` (or unmarked) function is passed around?

Comment (chriseth) [reply docs/types.rst:300]: You will get a type error

Comment (axic) [reply docs/types.rst:300]: Even if the `external` is removed from the Oracle example? So all the parameters default to `internal`.

**Comment by chriseth** [comment]
Return types are not part of the abi signature: Let's discuss that somewhere else.

**Comment by chriseth** [comment]
@pirapira discovered that 0 is actually not a valid jump destination, it was like that way before frontier.

File: docs/types.rst (hunk 20 above)
Comment (chriseth) [reply docs/types.rst:300]: Ah now I see what you mean. The default for a function type is "internal". Furthermore, the default for a function implementation is also "internal", unless it is defined at contract level scope, which makes it external and internal at the same time, i.e. public.

The type of a contract level function f is internal if you access it only as `f` and it is external if you access it as `this.f` - I will update the documentation to mention that.

**Comment by axic** [comment]
> >    discussion if that is the encoding we want (left aligned <function id><address>)
> 
> Will we also add conversion from function type to address since that seems natural? A recipient of a function type could decide not to call it if it refers to a third party address for example.

This reminds me: it should be right aligned (left padded) to support conversion by masking (and not shifting).

That would also change its output type from `bytes24` to `uint192`.

**Comment by chriseth** [comment]
@axic the main reason I changed it to `bytes24` was that it is much closer to a `bytes24` than it is to `uint192`. Furthermore, conversion by masking does not help much, you always have to shift, too.

**Comment by axic** [comment]
As an example, if it is right aligned: `Test(0xca35b7d915458ef540ade6068dfe2f44e8fa733c).confirm(bytes32)` becomes: `00000000797af627ca35b7d915458ef540ade6068dfe2f44e8fa733c`.

Typecasting that to `address` (`uint160`) should gives us the address.

**Comment by chriseth** [comment]
Simple typecasting to address is exactly one danger I would like to avoid. How the higher order bits are handled for address is unspecified, so if you pass an external function to a function that assumes an address, it will perform the conversion without warning and just ignore the function signature part.

**Comment by axic** [comment]
What is unspecified about higher order bits for addresses? Also the conversion from function to address should not be implicit.

**Comment by chriseth** [comment]
@axic I'm talking about using an external function type in the call data to a contract that expects address.

**Comment by chriseth** [comment]
Proposal by @axic: encode functions as left-aligned `<address><function id>` because the address is the more higher-level construct.

**Comment by chriseth** [comment]
Discussion about the ABI resulted in:
- we will use `<address><function id>` (left aligned) as encoding type, potentially using the zero bytes after that as runtime type information at a later point in time
- the ABI type will be `function` (new type name). We do not encode the parameter types because the ABI is only used to disambiguate overloaded functions and specify the encoding. Encoding the full type information of the function is too detailed - we also do not do that for a lot of other things (Contract, for example).

**Comment by chriseth** [comment]
@axic please comment on the latest changes.

File: libsolidity/interface/CompilerStack.cpp [hunk 21]
```diff
@@ -590,9 +591,19 @@ void CompilerStack::compileContract(
...
+		Compiler cloneCompiler(_optimize, _runs);
+		cloneCompiler.compileClone(_contract, _compiledContracts);
+		compiledContract.cloneObject = cloneCompiler.assembledObject();
+	}
+	catch (eth::AssemblyException const& _e)
+	{
+		// In some cases (if the constructor requests a runtime function), it is not
+		// possible to compile the clone.
+
+		// TODO: Report error / warning
```
Comment (axic) [inline libsolidity/interface/CompilerStack.cpp:605]: Why are we hiding the error here?

**Comment by axic** [comment]
We should have a test in `SolidityABIJSON` (and maybe in `ASTJSON` too).

File: docs/types.rst [hunk 22]
```diff
@@ -266,6 +266,127 @@ to and from all integer types but implicit conversion is not allowed.
...
+There are two ways to access a function in the current contract: Either directly
+by its name, ``f``, or using ``this.f``. The former will result in an internal
+function, the latter in an external function.
+
+If a function type variable is not initialized, calling it will result
+in an exception. The same happens if you call a function after using ``delete``
+on it.
+
+If external function types are used outside of the context of Solidity,
+they are converted into the ``bytes24`` type.
```
Comment (axic) [inline docs/types.rst]: Shouldn't this be `function` now instead of `bytes24`?

File: libsolidity/interface/CompilerStack.cpp (hunk 21 above)
Comment (chriseth) [reply libsolidity/interface/CompilerStack.cpp:605]: There is no easy way to report it (at this point, only throwing an exception would work, but this is exactly what we want to avoid) and clones are an undocumented feature anyway.

**Comment by chriseth** [comment]
Updated documentation and added test.

File: docs/types.rst [hunk 23]
```diff
@@ -266,6 +266,128 @@ to and from all integer types but implicit conversion is not allowed.
...
     }
 
+.. index:: ! function type, ! type; function
+
+.. _function_types:
+
+Function Types
+--------------
+
+Function types can be used to assign functions to variables and passing them
```
Comment (pirapira) [inline docs/types.rst]: `passing them to` and `return them from` do not balance.  I think it's either `pass them to and return them from` or `passing them to and returning them from`.

**Comment by chriseth** [comment]
@axic @pirapira ready to merge?

File: libsolidity/ast/Types.cpp [hunk 24]
```diff
@@ -1699,7 +1715,7 @@ TypePointer TupleType::closestTemporaryType(TypePointer const& _targetType) cons
 FunctionType::FunctionType(FunctionDefinition const& _function, bool _isInternal):
 	m_location(_isInternal ? Location::Internal : Location::External),
 	m_isConstant(_function.isDeclaredConst()),
-	m_isPayable(_function.isPayable()),
+	m_isPayable(_isInternal ? false : _function.isPayable()),
```
Comment (pirapira) [inline libsolidity/ast/Types.cpp:1723]: Can we invoke an assertion failure if a function is internal and payable at the same time?

File: libsolidity/ast/Types.cpp [hunk 25]
```diff
@@ -1800,6 +1816,33 @@ FunctionType::FunctionType(EventDefinition const& _event):
 	swap(paramNames, m_parameterNames);
 }
 
+FunctionType::FunctionType(FunctionTypeName const& _typeName):
+	m_location(_typeName.visibility() == VariableDeclaration::Visibility::External ? Location::External : Location::Internal),
+	m_isConstant(_typeName.isDeclaredConst()),
+	m_isPayable(_typeName.isPayable())
+{
+	if (_typeName.isPayable())
+		solAssert(m_location == Location::External, "Internal payable function type used.");
```
Comment (pirapira) [inline libsolidity/ast/Types.cpp:1831]: Can we also assert that one cannot have `isPayable` and `isConstant` at the same time?

File: libsolidity/ast/Types.cpp [hunk 26]
```diff
@@ -1800,6 +1816,33 @@ FunctionType::FunctionType(EventDefinition const& _event):
...
+			m_location != Location::External || t->annotation().type->canBeUsedExternally(false),
+			"Internal type used as parameter for external function."
+		);
+		m_parameterTypes.push_back(t->annotation().type);
+	}
+	for (auto const& t: _typeName.returnParameterTypes())
+	{
+		solAssert(t->annotation().type, "Type not set for return parameter.");
+		solAssert(
+			m_location != Location::External || t->annotation().type->canBeUsedExternally(false),
```
Comment (pirapira) [inline libsolidity/ast/Types.cpp]: The assertion reads strange to me.  Maybe `m_location == External /* then it's ok */ || ...`

File: libsolidity/ast/Types.cpp [hunk 27]
```diff
@@ -1875,22 +1918,65 @@ bool FunctionType::operator==(Type const& _other) const
...
 		name += (*it)->toString(_short) + (it + 1 == m_parameterTypes.end() ? "" : ",");
-	name += ") returns (";
+	name += ") ";
+	if (m_isConstant)
+		name += "constant ";
+	if (m_isPayable)
+		name += "payable ";
+	if (m_location == Location::External)
+		name += "external ";
+	name += "returns (";
```
Comment (pirapira) [inline libsolidity/ast/Types.cpp]: The `returns` clause should not exist when the number of return values is zero.  (So I read somewhere.)

File: libevmasm/Assembly.cpp [hunk 28]
```diff
@@ -75,17 +79,17 @@ string Assembly::out() const
 	return ret.str();
 }
 
-unsigned Assembly::bytesRequired() const
+unsigned Assembly::bytesRequired(unsigned subTagSize) const
 {
-	for (unsigned br = 1;; ++br)
+	for (unsigned tagSize = subTagSize;; ++tagSize)
```
Comment (axic) [inline libevmasm/Assembly.cpp]: It looks weird that there's no spacing between the two semicolons.

File: libsolidity/codegen/CompilerUtils.cpp [hunk 29]
```diff
@@ -133,6 +133,16 @@ void CompilerUtils::storeInMemoryDynamic(Type const& _type, bool _padToWordBound
 			m_context << u256(str->value().size());
 		m_context << Instruction::ADD;
 	}
+	else if (
+		_type.category() == Type::Category::Function &&
+		dynamic_cast<FunctionType const&>(_type).location() == FunctionType::Location::External
+	)
+	{
+		solAssert(_padToWordBoundaries, "Non-padded store for function not implemented.");
```
Comment (axic) [inline libsolidity/codegen/CompilerUtils.cpp]: Can this be moved to the `UnimplementedFeature` exception?

File: libsolidity/ast/Types.cpp (hunk 24 above)
Comment (chriseth) [reply libsolidity/ast/Types.cpp:1723]: I will try.

File: libsolidity/ast/Types.cpp (hunk 25 above)
Comment (chriseth) [reply libsolidity/ast/Types.cpp:1831]: That's why I really would like to combine those flags into a single enum or at least a single class that has internal checks for these restrictions.

File: libsolidity/ast/Types.cpp (hunk 26 above)
Comment (chriseth) [reply libsolidity/ast/Types.cpp]: This is a logical implication, but I will rewrite it using `if`.

File: libsolidity/ast/Types.cpp (hunk 24 above)
Comment (chriseth) [reply libsolidity/ast/Types.cpp:1723]: This is the constructor that creates a function type out of a function definition. If the function is `public`, it can be payable, and the resulting internal function type will not be payable.

**Comment by chriseth** [comment]
Updated.

**Review by pirapira — APPROVED** [review]
I think this one is ready to merge.
